packages/mui-system/src/createStyled.js

```diff
@@ -34,11 +34,18 @@ const getStyleOverrides = (name, theme) => {
 };
 
 const transformVariants = (variants) => {
+  let numOfCallbacks = 0;
   const variantsStyles = {};
 
   if (variants) {
     variants.forEach((definition) => {
-      const key = propsToClassKey(definition.props);
+      let key = '';
+      if (typeof definition.props === 'function') {
+        key = `callback${numOfCallbacks}`;
+        numOfCallbacks += 1;
```

Comment:
It should probably work, but for debugging purposes it may be confusing to see a callback 7 if you know there are only two callbacks defined. Happy to refactor if you feel strongly about it.

Reply:
Nope, my suggestion would clean up just one line so no need to refactor I guess.